### PR TITLE
Update setlogurl method for GoLocal support

### PR DESCRIPTION
**Pull request by jinquan2019**
Things to consider before you submit the PR:

* [ ] Has `CHANGELOG.md` been updated?
* [x] Are tests passing locally?
* [x] Are the files formatted correctly?
* [ ] Did you add unit tests?
* [ ] Did you check UI tests on the sample app? They are not executed on CI.
* [ ] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?

## Description

1. Change start API to take overridden log URL for One Collector
2. Add info message when setLogUrl is called and overridden.
3. Update setLogUrl method to change Ingestion endpoint immediately
    If appSecret exists => change App Center Ingestion endpoint.
    If appSecret doesn't exist => change One Collector Ingestion endpoint.

### CODE REVIEW

File: AppCenter/AppCenter/Internals/Channel/MSOneCollectorChannelDelegate.h [hunk 1]
```diff
@@ -14,6 +14,13 @@
  *
  * @return A `MSOneCollectorChannelDelegate` instance.
  */
-- (instancetype)initWithInstallId:(NSUUID *)installId;
+- (instancetype)initWithInstallId:(NSUUID *)installId oneCollectorBaseUrl:(NSString *)baseUrl;
```
Comment (jaelim-ms) [inline AppCenter/AppCenter/Internals/Channel/MSOneCollectorChannelDelegate.h]: `oneCollectorBaseUrl` -> `baseUrl`
Add `@param` doc for the parameter.

File: AppCenter/AppCenter/Internals/Channel/MSOneCollectorChannelDelegatePrivate.h [hunk 2]
```diff
@@ -24,6 +24,11 @@ extern NSString *const kMSLogNameRegex;
  */
 @property(nonatomic) MSOneCollectorIngestion *oneCollectorIngestion;
 
+/**
+ * Base Url for One Collector endpoint.
+ */
+@property(nonatomic) NSString *oneCollectorBaseUrl;
```
Comment (jaelim-ms) [inline AppCenter/AppCenter/Internals/Channel/MSOneCollectorChannelDelegatePrivate.h]: The class name already says "One Collector", rename to baseUrl.

File: AppCenter/AppCenter/MSAppCenter.m [hunk 3]
```diff
@@ -412,7 +412,13 @@ - (void)setLogUrl:(NSString *)logUrl {
   @synchronized(self) {
     _logUrl = logUrl;
     if (self.channelGroup) {
-      [self.channelGroup setLogUrl:logUrl];
+      if (!self.appSecret) {
```
Comment (jaelim-ms) [inline AppCenter/AppCenter/MSAppCenter.m]: Remove `!` and swap `if` and `else` block. It is unnecessary bit operation that you can remove.
```
if (!variable) {
  // Do A
} else {
  // Do B
}
```
can be rewritten to
```
if (variable) {
  // Do B
} else {
  // Do A
}
```

File: AppCenter/AppCenter/MSAppCenter.m [hunk 4]
```diff
@@ -554,8 +560,19 @@ - (void)applyPipelineEnabledState:(BOOL)isEnabled {
 - (void)initializeChannelGroup {
 
   // Construct channel group.
-  self.oneCollectorChannelDelegate =
-      self.oneCollectorChannelDelegate ?: [[MSOneCollectorChannelDelegate alloc] initWithInstallId:self.installId];
+  if (self.oneCollectorChannelDelegate) {
```
Comment (jaelim-ms) [inline AppCenter/AppCenter/MSAppCenter.m]: This implementation doesn't seem correct, Blocking the PR for now until I review this again. Sorry for delay.

Comment (ElektrojungeMS) [reply AppCenter/AppCenter/MSAppCenter.m]: hey, no worries!

Comment (jinquan2019) [reply AppCenter/AppCenter/MSAppCenter.m]: OK

**Comment by jaelim-ms** [comment]
So, the broken scenario for `setLogUrl` is
1. `setLogUrl` is called from library
2. Application started App Center w/ appSecret but there is no `setLogUrl` call made by an application
=> `logUrl` property in MSAppCenter will contain a custom endpoint for One Collector and it can be used for initializing App Center pipeline.

File: AppCenter/AppCenter/MSAppCenter.m (hunk 4 above)
Comment (jaelim-ms) [reply AppCenter/AppCenter/MSAppCenter.m]: This `if` statement doesn't seem required. You don't need to call `setLogUrl` here. Let's remove.

**Comment by jinquan2019** [comment]
> So, the broken scenario for `setLogUrl` is
> 
> 1. `setLogUrl` is called from library
> 2. Application started App Center w/ appSecret but there is no `setLogUrl` call made by an application
> 
> => `logUrl` property in MSAppCenter will contain a custom endpoint for One Collector and it can be used for initializing App Center pipeline.

I update the code to cover the scenario 2. 
In the 'initWithInstallId: baseUrl:' method of MSOneCollectorChannelDelegate class, it will check whether baseUrl is nil and assign a custom endpoint to it when baseUrl is nil. So if there is no 'setLogUrl' call, the 'logUrl' property in MSAppcenter will always be 'kMSAppCenterBaseUrl' and don't transfer the logUrl' property value to 'initWithInstallId: baseUrl:' method.

**Comment by jaelim-ms** [comment]
The scenario that I mentioned isn't separated two scenario. It is just steps for one scenario. Also, App Center has multiple endpoints for different environment so you can't actually do that way. Please ignore the scenario and revert the change.

File: Sasquatch/SasquatchObjC/AppDelegate.m [hunk 5]
```diff
@@ -94,6 +94,11 @@ - (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(
                    });
                  }];
   }
+    
+  NSString *logUrl = [[NSUserDefaults standardUserDefaults] stringForKey:kMSLogUrl];
+  if(logUrl){
```
Comment (jaelim-ms) [inline Sasquatch/SasquatchObjC/AppDelegate.m]: format

File: Sasquatch/Sasquatch/ViewControllers/MSMainViewController.swift [hunk 6]
```diff
@@ -144,6 +175,7 @@ class MSMainViewController: UITableViewController, AppCenterProtocol {
   @IBAction func dismissKeyboard(_ sender: UITextField!) {
     sender.resignFirstResponder()
   }
+
```
Comment (jaelim-ms) [inline Sasquatch/Sasquatch/ViewControllers/MSMainViewController.swift]: Remove this empty line.

File: Sasquatch/Sasquatch/ViewControllers/MSMainViewController.swift [hunk 7]
```diff
@@ -14,7 +14,7 @@ class MSMainViewController: UITableViewController, AppCenterProtocol {
     
     static let allValues = [AppCenter, OneCollector, Both, None, Skip]
   }
-
+
```
Comment (jaelim-ms) [inline Sasquatch/Sasquatch/ViewControllers/MSMainViewController.swift]: I don't think this change is necessary.

Comment (jinquan2019) [reply Sasquatch/Sasquatch/ViewControllers/MSMainViewController.swift]: OK, Jason will revert the last changes and create new branch based on this branch

**Comment by jaelim-ms** [comment]
@jinquan2019  No unit tests found for the PR. There should be another PR coming for tests.

**Comment by jaelim-ms** [comment]
Please do not merge, we are converting this branch as a feature branch for next release.

**Comment by jinquan2019** [comment]
Sorry, I try to fix the conflicts when merge the develop branch, almost conflicts are fixed except the SasquatchMac.storyboard file. I create a new branch to fix it. Please review it and merge.

File: SasquatchMac/SasquatchMac/ViewControllers/AppCenterViewController.swift [hunk 8]
```diff
@@ -165,4 +169,41 @@ class AppCenterViewController : NSViewController, NSTextFieldDelegate, CLLocatio
...
+    alert.accessoryView = scrollView
+    alert.alertStyle = NSWarningAlertStyle
+    switch(alert.runModal()) {
+    case NSAlertFirstButtonReturn:
+      UserDefaults.standard.removeObject(forKey: kMSLogUrl)
+      appCenter.setLogUrl(prodLogUrl())
+      break
+    case NSAlertSecondButtonReturn:
+      let text = textView.string ?? ""
+      let logUrl = !text.isEmpty ? text : nil
```
Comment (annakocheshkova) [inline SasquatchMac/SasquatchMac/ViewControllers/AppCenterViewController.swift]: Are you sure setting log url to `nil` is ok? I would not set anything and roll back to previous url in such case.

Comment (annakocheshkova) [reply SasquatchMac/SasquatchMac/ViewControllers/AppCenterViewController.swift]: Ah nvm just saw the second PR :)

File: AppCenter/AppCenterTests/MSAppCenterTests.m [hunk 9]
```diff
@@ -11,11 +11,14 @@
 #import "MSAppCenterPrivate.h"
 #import "MSChannelGroupDefault.h"
 #import "MSDeviceTrackerPrivate.h"
+#import "MSHttpIngestion.h"
```
Comment (MatkovIvan) [inline AppCenter/AppCenterTests/MSAppCenterTests.m]: Redundant

File: AppCenter/AppCenterTests/MSAppCenterTests.m [hunk 10]
```diff
@@ -11,11 +11,14 @@
...
 #import "MSHttpIngestionPrivate.h"
 #import "MSMockSecondService.h"
 #import "MSMockService.h"
 #import "MSMockUserDefaults.h"
 #import "MSOneCollectorChannelDelegate.h"
```
Comment (MatkovIvan) [inline AppCenter/AppCenterTests/MSAppCenterTests.m]: Redundant

**Pull request merged**